File: Cholesky.m
% Usa la classe OSmemory per la misurazione della memoria
memoryFunc = OSmemory;

matrixNames = {'apache2.mat', 'cfd1.mat', 'cfd2.mat', 'ex15.mat', 'Flan_1565.mat', 'G3_circuit.mat', 'parabolic_fem.mat', 'shallow_water1.mat', 'StocF-1465.mat'};

% Inizializza una struttura per memorizzare i risultati
results = struct('File', {}, 'N', {}, 'Errore_Relativo', {}, 'Time', {}, 'Memory_Used', {}, 'Status', {});

for i = 1:length(matrixNames)
    % Usa fullfile per creare percorsi file indipendenti dal sistema operativo
    mtrx = load(fullfile(script_path, 'Matrix', matrixNames{i}));
    start_memory = memoryFunc.memory();
    disp(['Memoria iniziale per ', matrixNames{i}, ': ', num2str(start_memory)]);
    matrix = mtrx.Problem.A;

    % Creazione del vettore xe
    n = size(matrix, 1);
    xe = ones(n, 1);

    try
        tic;
        % Calcola b
        b = matrix * xe;
        % Decomposizione di Cholesky della matrice A
        R = chol(matrix);
        % Risolvi il sistema Ax = b
        % Risolvi per y nel sistema triangolare inferiore R' * y = b
        y = R' \ b;
        % Risolvi per x nel sistema triangolare superiore R * x = y
        x = R \ y;
        time = toc;

        % Misura la memoria finale
        final_memory = memoryFunc.memory();
        disp(['Memoria finale per ', matrixNames{i}, ': ', num2str(final_memory)]);

        % Calcola la memoria utilizzata
        memory_used = (final_memory - start_memory) / 1e6; % In MB

        % Controllo e gestione della memoria negativa
        if memory_used < 0
            % Se la memoria è negativa, aspetta un momento e ricontrolla
            pause(1);
            final_memory = memoryFunc.memory();
            memory_used = (final_memory - start_memory) / 1e6; % In MB

            % Se è ancora negativa, segna come errore
            if memory_used < 0
                error(['Memoria negativa rilevata per ', matrixNames{i}]);
            end
        end

        % Verifica l'errore
        errore_relativo = norm(x - xe, 2) / norm(xe, 2);

        % Salva i risultati nella struttura
        results(i).File = matrixNames{i};
        results(i).N = n;
        results(i).Errore_Relativo = errore_relativo;
        results(i).Time = time;
        results(i).Memory_Used = memory_used;
        results(i).Status = 'Success';

        % Pulisci le variabili per liberare memoria
        clear mtrx matrix x y R b
    catch ME
        % Se c'è un errore (es. memoria esaurita), salva le informazioni rilevanti
        results(i).File = matrixNames{i};
        results(i).N = n;
        results(i).Errore_Relativo = NaN;
        results(i).Time = NaN;
        results(i).Memory_Used = NaN;
        results(i).Status = ['Error: ', ME.message];
    end

    % Forza la garbage collection
    pause(0.1);
    java.lang.System.gc()
end

% Informazioni di sistema
system_info = struct('Language', 'MATLAB', 'Operating_System', computer);

% Risultati finali
final_results = struct('System_Info', system_info, 'Matrix_Results', results);

% Leggi il file JSON esistente
json_file = fullfile(script_path, 'results.json');
if exist(json_file, 'file')
    fid = fopen(json_file, 'r');
    raw = fread(fid, inf, 'uint8=>char')';
    fclose(fid);
    existing_data = jsondecode(raw);
else
    existing_data = struct();
end

% Aggiungi i risultati di MATLAB mantenendo i dati esistenti
if ispc
    existing_data.Windows_MATLAB = final_results;
else
    existing_data.Linux_MATLAB = final_results;
end

% Converti la struttura in JSON e salva nel file
jsonStr = jsonencode(existing_data);
fid = fopen(json_file, 'w');
if fid == -1
    error('Impossibile aprire il file per la scrittura.');
end
fwrite(fid, jsonStr, 'char');
fclose(fid);

File: cholesky.py
import time
import scipy.io
import numpy as np
from scipy.sparse import csc_matrix
from scipy.sparse.linalg import eigsh
from numpy.linalg import norm
import sksparse.cholmod as cholmod
import psutil
import os
import json
import platform


def get_memory_usage():
    process = psutil.Process()
    memoria = process.memory_info().rss / (1024 * 1024)

    """
    if os.name == 'posix':  # Linux
        memoria = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024
        print(memoria_finale)
    elif os.name == 'nt':  # Windows
        process = psutil.Process()
        memoria = process.memory_info().rss / (1024 * 1024)
    """
    return memoria


def is_symmetric(mtrx):
    #  verifica se il numero di elementi diversi è zero.
    if (mtrx != mtrx.T).nnz == 0:
        print(f"Matrice è simmetrica")
        return True
    else:
        print(f"Matrice non è simmetrica")
        return False


def is_positive_definite(mtrx):
    try:
        eigenvalue, _ = eigsh(mtrx, k=1, which='SM')
        if eigenvalue > 0:
            print(f"Matrice definita positiva")
            return True
        else:
            print(f"Matrice non definita positiva")
            return False
    except:
        print(f"Matrice non definita positiva")
        return False


def solution(matrix):
    n = matrix.shape[0]
    xe = np.ones(n)

    start_time = time.time()

    b = matrix @ xe
    try:
        factor = cholmod.cholesky(matrix)
        x = factor(b)
        computation_time = time.time() - start_time

        errore_relativo = norm(x - xe) / norm(xe)
    except cholmod.CholmodError:
        errore_relativo = None
        computation_time = None

    return errore_relativo, computation_time


def process_matrix(matrix_name):
    json_structure = {
        'File': matrix_name,
        'Errore_Relativo': None,
        'Time': None,
        'Memory_Used': None,
        'Status': 'Failed'
    }

    try:
        data = scipy.io.loadmat(f'Matrix/{matrix_name}')
        A = csc_matrix(data['Problem'][0, 0]['A'])

        memory_after_load = get_memory_usage()

        if is_symmetric(A):
            errore_relativo, mtrx_time = solution(A)
            memory_after_solution = get_memory_usage()

            json_structure['N'] = A.shape[0]
            json_structure['Errore_Relativo'] = float(errore_relativo)
            json_structure['Time'] = round(mtrx_time, 3)
            json_structure['Memory_Used'] = memory_after_solution - memory_after_load
            json_structure['Status'] = 'Success'
        else:
            json_structure['Status'] = 'Matrix not symmetric or not positive definite'

    except MemoryError:
        json_structure['Status'] = 'Memory Out of bound'
    except scipy.io.matlab.miobase.MatReadError:
        json_structure['Status'] = 'Error reading .mat file'
    except Exception as e:
        json_structure['Status'] = f'Error: {str(e)}'

    return json_structure


def load_existing_data():
    try:
        with open('results.json', 'r') as f:
            content = f.read()
            if content:
                return json.loads(content)
            else:
                return {}
    except FileNotFoundError:
        return {}
    except json.JSONDecodeError:
        print("Warning: The JSON file is not properly formatted. Starting with an empty dictionary.")
        return {}


def save_results(data):
    with open('results.json', 'w') as f:
        json.dump(data, f, indent=2)


if __name__ == '__main__':
    matrixNames = ['cfd1.mat', 'cfd2.mat', 'apache2.mat', 'ex15.mat',
                   'G3_circuit.mat', 'parabolic_fem.mat', 'shallow_water1.mat']

    # Carica i dati esistenti
    existing_data = load_existing_data()

    # Identifica il sistema operativo e il linguaggio corrente
    current_os = platform.system()
    current_language = 'Python'

    # Crea una chiave unica per questa combinazione di OS e linguaggio
    key = f"{current_os}_{current_language}"

    current_results = []

    for matrix in matrixNames:
        print(f"Processing {matrix} ...")
        try:
            json_result = process_matrix(matrix)
            current_results.append(json_result)
        except Exception as e:
            print(f"Error processing {matrix}: {str(e)}")
            current_results.append({
                'File': matrix,
                'Status': f'Error: {str(e)}'
            })

    # Aggiorna o aggiungi i risultati per questa combinazione di OS e linguaggio
    if key not in existing_data:
        existing_data[key] = {}

    existing_data[key]['System_Info'] = {
        'Operating_System': current_os,
        'Language': current_language
    }
    existing_data[key]['Matrix_Results'] = current_results

    # Salva i risultati aggiornati
    save_results(existing_data)

    print("Results saved to results.json")